File: C++/Creational_Patterns/AbstractFactory.cpp
#include <iostream>
#include <vector>

class Shape
{
public:
	Shape() {}
	virtual void draw() = 0;
};

class Circle : public Shape
{
public:
	void draw()
	{
		std::cout << "Circle " << ": draw" << std::endl;
	}
};

class Ellipse : public Shape
{
public:
	void draw()
	{
		std::cout << "Ellipse " << ": draw" << std::endl;
	}
};

class Square : public Shape
{
public:
	void draw()
	{
		std::cout << "Square " << ": draw" << std::endl;
	}
};

class Rectangle : public Shape
{
public:
	void draw()
	{
		std::cout << "Rectangle " << ": draw" << std::endl;
	}
};

class ShapeFactory
{
public:
	virtual Shape* createCurvedInstance() = 0;
	virtual Shape* createStraightInstance() = 0;
};

class SimpleShapeFactory : public ShapeFactory
{
	Shape* createCurvedInstance()
	{
		return new Circle;
	}

	Shape* createStraightInstance()
	{
		return new Square;
	}
};

class ComplexShapeFactory : public ShapeFactory
{
public:
	Shape* createCurvedInstance()
	{
		return new Ellipse;
	}

	Shape* createStraightInstance()
	{
		return new Rectangle;
	}
};

int main()
{
	std::cout << ":: Abstract Factory ::\n" << std::endl;

	std::vector<Shape*> shapes;

	ShapeFactory* simpleFactory = new SimpleShapeFactory;
	shapes.push_back(simpleFactory->createCurvedInstance());    // new Circle;
	shapes.push_back(simpleFactory->createStraightInstance());  // new Square;

	ShapeFactory* complexFactory = new ComplexShapeFactory;
	shapes.push_back(complexFactory->createCurvedInstance());   // new Ellipse;
	shapes.push_back(complexFactory->createStraightInstance()); // new Rectangle;

	for (auto& shape : shapes)
	{
		shape->draw();
	}

	system("PAUSE");

	return 0;
}

File: creational/javascript/AbstractFactory.js
'use strict';

class Shape {
  constructor() {
    if (this.constructor === Shape) {
      throw new TypeError('Cannot construct Abstract instances directly');
    }

    if (this.draw === Shape.prototype.draw) {
      throw new TypeError('Must override method "draw()"');
    }
  }

  draw() {
    throw new TypeError('Do not call abstract method "draw()" from child.');
  }
}

class Circle extends Shape {
  constructor() {
    super();
  }

  draw() {
    console.log('Circle: draw');
  }
}

class Ellipse extends Shape {
  constructor() {
    super();
  }

  draw() {
    console.log('Ellipse: draw');
  }
}

class Square extends Shape {
  constructor() {
    super();
  }

  draw() {
    console.log('Square: draw');
  }
}

class Rectangle extends Shape {
  constructor() {
    super();
  }

  draw() {
    console.log('Rectangle: draw');
  }
}

class ShapeFactory {
  contructor() {
    if (this.constructor === ShapeFactory) {
      throw new TypeError('Cannot construct Abstract instances directly');
    }

    if (this.createCurvedInstance === ShapeFactory.prototype.createCurvedInstance) {
      throw new TypeError('Must override method "createCurvedInstance()"');
    }

    if (this.createStraightInstance === ShapeFactory.prototype.createStraightInstance) {
      throw new TypeError('Must define method "createStraightInstance()"');
    }
  }

  createCurvedInstance() {
    throw new TypeError('Do not call abstract method "createCurvedInstance()" from child.');
  }

  createStraightInstance() {
    throw new TypeError('Do not call abstract method "createStraightInstance()" from child.');
  }
}

class SimpleShapeFactory extends ShapeFactory {
  constructor() {
    super();
  }

  createCurvedInstance() {
    return new Circle();
  }

  createStraightInstance() {
    return new Square();
  }
}

class ComplexShapeFactory extends ShapeFactory {
  constructor() {
    super();
  }

  createCurvedInstance() {
    return new Ellipse();
  }

  createStraightInstance() {
    return new Rectangle();
  }
}

console.log(':: Abstract Factory ::\n');

let shapes = [];

let simpleFactory = new SimpleShapeFactory();
shapes.push(simpleFactory.createCurvedInstance());
shapes.push(simpleFactory.createStraightInstance());

let complexFactory = new ComplexShapeFactory();
shapes.push(complexFactory.createCurvedInstance());
shapes.push(complexFactory.createStraightInstance());

shapes.forEach((shape) => {
  shape.draw();
});
